internal/connectors/plugins/public/qonto/accounts.go

```diff
@@ -0,0 +1,104 @@
+package qonto
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/formancehq/go-libs/v3/pointer"
+	"github.com/formancehq/payments/internal/connectors/plugins/currency"
+	"github.com/formancehq/payments/internal/connectors/plugins/public/qonto/client"
+	"time"
+
+	"github.com/formancehq/payments/internal/models"
+)
+
+type accountsState struct {
+	LastUpdatedAt time.Time `json:"lastUpdatedAt"`
+}
+
+func (p *Plugin) fetchNextAccounts(ctx context.Context, req models.FetchNextAccountsRequest) (models.FetchNextAccountsResponse, error) {
+	var oldState accountsState
+	if req.State != nil {
+		if err := json.Unmarshal(req.State, &oldState); err != nil {
+			return models.FetchNextAccountsResponse{}, err
+		}
+	} else {
+		oldState = accountsState{}
+	}
+
+	newState := accountsState{
+		LastUpdatedAt: oldState.LastUpdatedAt,
+	}
+
+	var accounts []models.PSPAccount
+
+	organization, err := p.client.GetOrganization(ctx)
+	if err != nil {
+		return models.FetchNextAccountsResponse{}, err
+	}
+	accounts, newState.LastUpdatedAt, err = fillAccounts(organization.BankAccounts, accounts, oldState)
+	if err != nil {
+		return models.FetchNextAccountsResponse{}, err
+	}
+
+	payload, err := json.Marshal(newState)
+	if err != nil {
+		return models.FetchNextAccountsResponse{}, err
+	}
+
+	/*
+		TODO
+		Need to check the behaviour of the framework:
+		* We are always keeping the "lastUpdatedAt", I believe this gets stored to ensure we don't save/query multiple time
+		the same account -- is that true? Note that the API always return all accounts, so we can't use that for filtering
+		on the API side.
+		* The PSPAccount only has a notion of CreatedAt, and the API has only "updatedAt". It might not be a big deal
+		if the framework can support upsert by reference?
+	*/
+	return models.FetchNextAccountsResponse{
+		Accounts: accounts,
+		NewState: payload,
+	}, nil
+}
+
+func fillAccounts(
+	bankAccounts []client.QontoBankAccount,
+	accounts []models.PSPAccount,
+	oldState accountsState,
+) ([]models.PSPAccount, time.Time, error) {
+	newestUpdatedAt := time.Time{}
+	for _, bankAccount := range bankAccounts {
+
+		// TODO check date format (particularly timezone -- the dates passed in are in UTC, we should save that info)
+		updatedAt, err := time.Parse("2006-01-02T15:04:05.999Z", bankAccount.UpdatedAt)
+		if err != nil {
+			return nil, newestUpdatedAt, err
+		}
+
+		// Ignore accounts that have already been processed
+		switch updatedAt.Compare(oldState.LastUpdatedAt) {
+		case -1, 0:
+			// Account already ingested, skip
+			continue
```

Comment:
I'm not sure this is meaningful unless you sort the accounts by UpdatedAt and ensure the oldest UpdatedAt is stored in state.

Imagine you have a page size of 5 and there are 6 accounts in total:
| id | updated_at |
| --- | --- |
| 1 | 2025-01-30 |
| 2 | 2025-03-30 |
| 3 | 2025-02-05 |
| 4 | 2025-03-22 |
| 5 | 2025-04-15 |
| 6 | 2025-01-29 |

With your implementation we'll ingest the first 5 accounts, then we save `2025-04-15` in the state. When we check again, we fail to ingest 6 because that account hasn't been updated on Qonto side recently.

Reply:
yeah good call 😂 I'm fixing that